addons/mail/models/mail_channel.py

```diff
@@ -723,30 +723,30 @@ def _execute_channel_pin(self, pinned=False):
         if channel_partners:
             channel_partners.write({'is_pinned': pinned})
 
-    def channel_seen(self):
+    def channel_seen(self, last_message_id):
         self.ensure_one()
-        if self.channel_message_ids.ids:
-            last_message_id = self.channel_message_ids.ids[0] # zero is the index of the last message
-            channel_partner = self.env['mail.channel.partner'].search([('channel_id', 'in', self.ids), ('partner_id', '=', self.env.user.partner_id.id)], limit=1)
-            if channel_partner.seen_message_id.id == last_message_id:
-                # last message seen by user is already up-to-date
-                return
-            channel_partner.write({
-                'seen_message_id': last_message_id,
-                'fetched_message_id': last_message_id,
-            })
-            data = {
-                'id': channel_partner.id,
-                'info': 'channel_seen',
-                'last_message_id': last_message_id,
-                'partner_id': self.env.user.partner_id.id,
-            }
-            if self.channel_type == 'chat':
-                self.env['bus.bus'].sendmany([[(self._cr.dbname, 'mail.channel', self.id), data]])
-            else:
-                data['channel_id'] = self.id
-                self.env['bus.bus'].sendone((self._cr.dbname, 'res.partner', self.env.user.partner_id.id), data)
-            return last_message_id
+        last_message = self.env['mail.message'].browse(last_message_id)
+        if self not in last_message.channel_ids:
+            return  # message is not in current channel
```

Comment:
Now that I think about it, the existence of the message does not actually matter (in case it was deleted in the meantime).

I think the proper way to select the `last_message` would be to `search` for the most recent message in the current channel that has id < provided id.

If we want to keep the support for no param, we could fallback to not add "id < provided id" in search domain when no id is provided, so that it gets the last message like before (but with a query with 1 result instead of fetching them all through related field).